Fix an assert failure with a funclet in a swifttailcc function.

hjyamauchi opened the pull request:
The failure happens in the livedebugvalues pass.

hjyamauchi commented:
The assert failure dump looks like

```# | Assertion failed: (Subtarget.isCallingConvWin64(MF.getFunction().getCallingConv()) ) && "Funclets should only be present on Win64 or swifttailcc", file C:\Users\hiroshi\llvm-project\llvm\lib\Target\AArch64\AArch64FrameLowering.cpp, line 2596
# | PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
# | Stack dump:
# | 0.  Program arguments: c:\\users\\hiroshi\\llvm-project\\build\\bin\\llc.exe -o -
# | 1.  Running pass 'Function Pass Manager' on module '<stdin>'.
# | 2.  Running pass 'Live DEBUG_VALUE analysis' on function '@foo'
# | Exception Code: 0x80000003
# |  #0 0x00007ff7c1f8a3f5 HandleAbort C:\Users\hiroshi\llvm-project\llvm\lib\Support\Windows\Signals.inc:424:0
# |  #1 0x00007ffe39b3e5f5 (C:\Windows\System32\ucrtbase.dll+0x7e5f5)
# |  #2 0x00007ffe39b3f601 (C:\Windows\System32\ucrtbase.dll+0x7f601)
# |  #3 0x00007ffe39b40fae (C:\Windows\System32\ucrtbase.dll+0x80fae)
# |  #4 0x00007ffe39b411f1 (C:\Windows\System32\ucrtbase.dll+0x811f1)
# |  #5 0x00007ff7befa0f86 llvm::AArch64FrameLowering::resolveFrameOffsetReference(class llvm::MachineFunction const &, __int64, bool, bool, class llvm::Register &, bool, bool) const C:\Users\hiroshi\llvm-project\llvm\lib\Target\AArch64\AArch64FrameLowering.cpp:2598:0
# |  #6 0x00007ff7befa0b5d llvm::AArch64FrameLowering::resolveFrameIndexReference(class llvm::MachineFunction const &, int, class llvm::Register &, bool, bool) const C:\Users\hiroshi\llvm-project\llvm\lib\Target\AArch64\AArch64FrameLowering.cpp:2522:0
# |  #7 0x00007ff7bef9a131 llvm::AArch64FrameLowering::getFrameIndexReference(class llvm::MachineFunction const &, int, class llvm::Register &) const C:\Users\hiroshi\llvm-project\llvm\lib\Target\AArch64\AArch64FrameLowering.cpp:2474:0
# |  #8 0x00007ff7c1518934 `anonymous namespace'::VarLocBasedLDV::extractSpillBaseRegAndOffset C:\Users\hiroshi\llvm-project\llvm\lib\CodeGen\LiveDebugValues\VarLocBasedImpl.cpp:1303:0
# |  #9 0x00007ff7c1525b38 `anonymous-namespace'::VarLocBasedLDV::OpenRangesSet::getSpillVarLocs C:\Users\hiroshi\llvm-project\llvm\lib\CodeGen\LiveDebugValues\VarLocBasedImpl.cpp:962:0
# | #10 0x00007ff7c1525b38 `anonymous namespace'::VarLocBasedLDV::transferSpillOrRestoreInst C:\Users\hiroshi\llvm-project\llvm\lib\CodeGen\LiveDebugValues\VarLocBasedImpl.cpp:1761:0
# | #11 0x00007ff7c151095d llvm::PointerIntPairInfo<llvm::ilist_node_base<1> *,1,llvm::PointerLikeTypeTraits<llvm::ilist_node_base<1> *> >::getInt C:\Users\hiroshi\llvm-project\llvm\include\llvm\ADT\PointerIntPair.h:196:0
# | #12 0x00007ff7c151095d llvm::PointerIntPair<llvm::ilist_node_base<1> *,1,unsigned int,llvm::PointerLikeTypeTraits<llvm::ilist_node_base<1> *>,llvm::PointerIntPairInfo<llvm::ilist_node_base<1> *,1,llvm::PointerLikeTypeTraits<llvm::ilist_node_base<1> *> > >::getInt C:\Users\hiroshi\llvm-project\llvm\include\llvm\ADT\PointerIntPair.h:96:0
# | #13 0x00007ff7c151095d llvm::ilist_node_base<1>::isSentinel C:\Users\hiroshi\llvm-project\llvm\include\llvm\ADT\ilist_node_base.h:45:0
# | #14 0x00007ff7c151095d llvm::ilist_node_impl<llvm::ilist_detail::node_options<llvm::MachineInstr,1,1,void,0> >::isSentinel C:\Users\hiroshi\llvm-project\llvm\include\llvm\ADT\ilist_node.h:130:0
# | #15 0x00007ff7c151095d llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::MachineInstr,1,1,void,0>,0,0>::isEnd C:\Users\hiroshi\llvm-project\llvm\include\llvm\ADT\ilist_iterator.h:175:0
# | #16 0x00007ff7c151095d llvm::MachineInstrBundleIteratorHelper<0>::getBundleFinal C:\Users\hiroshi\llvm-project\llvm\include\llvm\CodeGen\MachineInstrBundleIterator.h:62:0
# | #17 0x00007ff7c151095d llvm::MachineInstrBundleIteratorHelper<0>::increment C:\Users\hiroshi\llvm-project\llvm\include\llvm\CodeGen\MachineInstrBundleIterator.h:70:0
# | #18 0x00007ff7c151095d llvm::MachineInstrBundleIterator<llvm::MachineInstr,0>::operator++ C:\Users\hiroshi\llvm-project\llvm\include\llvm\CodeGen\MachineInstrBundleIterator.h:260:0
# | #19 0x00007ff7c151095d `anonymous namespace'::VarLocBasedLDV::ExtendRanges C:\Users\hiroshi\llvm-project\llvm\lib\CodeGen\LiveDebugValues\VarLocBasedImpl.cpp:2344:0
# | #20 0x00007ff7c12cf58b `anonymous namespace'::LiveDebugValues::runOnMachineFunction C:\Users\hiroshi\llvm-project\llvm\lib\CodeGen\LiveDebugValues\LiveDebugValues.cpp:129:0
# | #21 0x00007ff7c108f822 llvm::MachineFunctionPass::runOnFunction(class llvm::Function &) C:\Users\hiroshi\llvm-project\llvm\lib\CodeGen\MachineFunctionPass.cpp:95:0# | #22 0x00007ff7c15ae120 llvm::FPPassManager::runOnFunction(class llvm::Function &) C:\Users\hiroshi\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1443:0
```

compnerd commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2590,9 +2590,11 @@ StackOffset AArch64FrameLowering::resolveFrameOffsetReference(
-            "Funclets should only be present on Win64");
+            (Subtarget.isCallingConvWin64(MF.getFunction().getCallingConv()) ||
+             MF.getFunction().getCallingConv() == CallingConv::SwiftTail) &&
+            "Funclets should only be present on Win64 or swifttailcc");
```
What happens if I have a `swiftcc` function? Is there a reason that loosening the assertion like this is appropriate?

hjyamauchi replied:
swiftcc is already covered by isCallingConvWin64. I think this is checking that the particular cc uses a frame pointer, which swifttailcc does, and being relatively a recent addition, it wasn't covered by this code.

compnerd replied:
Ah, okay!  That I think is relevant details that should be in the commit message.

hjyamauchi replied:
Done

compnerd commented:
CC: @mstorsjo

mstorsjo commented on llvm/test/CodeGen/AArch64/swifttail-funclet.ll:
```diff
@@ -0,0 +1,125 @@
+!llvm.dbg.cu = !{!0, !11}
+!llvm.module.flags = !{!87}
+
+!0 = distinct !DICompileUnit(language: DW_LANG_Swift, file: !1, producer: "Swift version 5.11-dev", isOptimized: true, flags: "-private-discriminator _0FA2C3DB08FF346F3CA00B2EC660E1DF -enable-experimental-cxx-interop", runtimeVersion: 5, emissionKind: FullDebug, globals: !2, imports: !10, sysroot: "S:\\Windows.sdk", sdk: "Windows.sdk")
```
Is it possible to regenerate the input IR for this testcase, without all the extra debug data, which I presume isn't relevant for the thing being tested?

hjyamauchi replied:
The debug info is relevant here because the failing pass is `Live DEBUG_VALUE analysis` or `livedebugvalues`. It doesn't crash without it.

mstorsjo commented:
> CC: @mstorsjo

Thanks for looping me in; I don't have any opinion on the fix itself though.

compnerd reviewed (approved):
Please add the additional details to the commit message before pushing to main.

hjyamauchi commented:
> Please add the additional details to the commit message before pushing to main.

Done

hjyamauchi committed:
Fix an assert failure with a funclet in a swifttailcc function.

The failure happens in the livedebugvalues pass.

swiftcc is already covered by isCallingConvWin64. I believe that the
assert is checking that the particular calling convention uses a frame
pointer, which swifttailcc does, and being relatively a recent
addition, it wasn't covered by this code.

hjyamauchi commented:
PTAL @compnerd. The previous reversion turned out to be incorrect. I think the latest revision should work. I believe that SwiftTailCC should be treated as Win64 as seen in [the X86 equivalent](https://github.com/llvm/llvm-project/blob/db4ea21dfde307f4fc873a6fe56791cd7ae3f0a1/llvm/lib/Target/X86/X86Subtarget.h#L357) and the fixed object size needs to be adjusted to account for the tail call reserved stack for SwiftTailCC.

compnerd reviewed (approved):
Thank you! This seems much more reasonable.
